## Parte 2 - ABT
---
<img align="center" width="150" src="https://logodownload.org/wp-content/uploads/2017/05/ifood-logo-0.png">



### Acessando as bases

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

ValueError: mount failed

In [ ]:
import os
import numpy as np
import pandas as pd

## Lendos os Dados

In [ ]:
df_orders = pd.read_csv('/content/drive/MyDrive/projeto-ia-datasets/ifood/df_orders.csv')

In [ ]:
df_orders.shape

(492138, 43)

In [ ]:
df_orders.head()

order_id  order_number  \
0  630e2af0-b456-4b3a-b964-4d66ce5cc5df    1290139943   
1  66f8163d-f081-4fbd-adb8-cfbbff9213f7    1141271215   
2  e8c55557-81d4-4159-bd2b-c144e5fbabe3     694742752   
3  34a60e0d-0042-49af-9aa8-4dc526a998c2     679245732   
4  998f6d97-f884-40ba-a7a3-e9691930a642    1052187003   

      order_timestamp_local     order_shift    last_status_date_local  \
0  2019-12-11T13:22:06.497Z   weekday lunch  2019-12-11T15:22:39.156Z   
1  2019-11-06T17:57:17.967Z  weekday dinner  2019-11-06T18:18:48.680Z   
2  2019-06-11T20:42:32.058Z  weekday dinner  2019-06-11T22:42:54.005Z   
3  2019-06-05T19:16:42.111Z  weekday dinner  2019-06-05T21:17:13.264Z   
4  2019-10-15T21:01:23.973Z  weekday dinner  2019-10-15T23:01:41.146Z   

   order_total  credit  paid_amount delivery_type  scheduled  ...  \
0        28.90    14.0         24.9      DELIVERY      False  ...   
1        56.35     7.9         50.4      DELIVERY      False  ...   
2        53.80     4.0         49.8      DELIVERY      False  ...   
3        58.90     4.0         54.9      DELIVERY      False  ...   
4        54.80    14.0         50.8      DELIVERY      False  ...   

  distance_merchant_customer promo_is_promotion normal_items_quantity  \
0                2722.603906                0.0                   2.0   
1                1108.823183                1.0                   0.0   
2                1642.258522                0.0                   2.0   
3                 441.677507                0.0                   1.0   
4                4398.730930                1.0                   0.0   

  promo_items_quantity order_lag_at_login order_lead_at_login  \
0                  0.0                2.0                 1.0   
1                  1.0                7.0                 1.0   
2                  0.0                5.0                 1.0   
3                  0.0                9.0                 2.0   
4                  2.0                3.0                 4.0   

           order_date_local valid_order                            session_id  \
0  2019-12-11T00:00:00.000Z           1  ccbcf9e2-3728-4458-a813-4031ae22e4bd   
1  2019-11-06T00:00:00.000Z           1  d9e94d88-ca88-4997-ad4c-62b9c0a35c1b   
2  2019-06-11T00:00:00.000Z           1  035b33cd-6ddf-4ee2-a887-7ba490bb9700   
3  2019-06-05T00:00:00.000Z           1  93f66506-a8ba-4c22-97ea-7489131958bb   
4  2019-10-15T00:00:00.000Z           1  515a402d-8b86-4e6d-9852-f02282736781   

                                         customer_id  
0  48b6d81407646ca61163f064a426ce38ac08bed2ce84f0...  
1  b3f74186880f5f9a79e0865ad8beab47e22b785b8b3bc0...  
2  c4f8999d4cb83cac8134b2051ffc611d5d660c049c3744...  
3  ac89a3eb35853705a09d54f1e5988ad876628b986045cd...  
4  fdef6bf0530b1f6507f175167aa442ade79b6e6fa085c9...  

[5 rows x 43 columns]

In [ ]:
df_orders.dtypes

order_id                                     object
order_number                                  int64
order_timestamp_local                        object
order_shift                                  object
last_status_date_local                       object
order_total                                 float64
credit                                      float64
paid_amount                                 float64
delivery_type                                object
scheduled                                      bool
scheduled_creation_date_local                object
device_app_version                           object
device_type                                  object
device_platform                              object
payment_method                               object
customer_state_label                         object
customer_city                                object
customer_district                            object
customer_centroid_id                         object
customer_has

In [ ]:
df_orders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 492138 entries, 0 to 492137
Data columns (total 43 columns):
 #   Column                                    Non-Null Count   Dtype  
---  ------                                    --------------   -----  
 0   order_id                                  492138 non-null  object 
 1   order_number                              492138 non-null  int64  
 2   order_timestamp_local                     492138 non-null  object 
 3   order_shift                               492138 non-null  object 
 4   last_status_date_local                    492138 non-null  object 
 5   order_total                               492138 non-null  float64
 6   credit                                    492138 non-null  float64
 7   paid_amount                               492138 non-null  float64
 8   delivery_type                             492138 non-null  object 
 9   scheduled                                 492138 non-null  bool   
 10  scheduled_creation_d

In [ ]:
df_orders.columns

Index(['order_id', 'order_number', 'order_timestamp_local', 'order_shift',
       'last_status_date_local', 'order_total', 'credit', 'paid_amount',
       'delivery_type', 'scheduled', 'scheduled_creation_date_local',
       'device_app_version', 'device_type', 'device_platform',
       'payment_method', 'customer_state_label', 'customer_city',
       'customer_district', 'customer_centroid_id', 'customer_has_plus',
       'customer_seg_status_last_month', 'customer_seg_recency_bucket',
       'customer_seg_frequency_bucket', 'customer_seg_merchant_offer_bucket',
       'customer_seg_benefits_sensitivity_bucket', 'customer_seg_marlin_tag',
       'customer_seg_gross_income_bucket', 'customer_seg_preferred_shift',
       'frn_id', 'merchant_city', 'merchant_district', 'merchant_centroid_id',
       'merchant_dish_type', 'distance_merchant_customer',
       'promo_is_promotion', 'normal_items_quantity', 'promo_items_quantity',
       'order_lag_at_login', 'order_lead_at_login', 'order_da

## Análise Exploratória

**`Q1`** Quantas linhas e colunas existem na base `df_orders`?

In [ ]:
df_orders.shape

(492138, 43)

**`Q2`** Verifique se todos os pedidos são unicos.

In [ ]:
df_orders['order_id'].nunique()

492138

**`Q3`** Qual o pedido mais antigo e o mais recente da base? Utilize a variável `order_timestamp_local`.

In [ ]:
df_orders["order_timestamp_local"].min()

'2019-06-01T00:00:15.369Z'

In [ ]:
df_orders["order_timestamp_local"].max()

'2019-12-31T23:31:59.972Z'

In [ ]:
df_orders["order_timestamp_local"].describe()

<ipython-input-16-1ef113766b76>:1: FutureWarning: Treating datetime data as categorical rather than numeric in `.describe` is deprecated and will be removed in a future version of pandas. Specify `datetime_is_numeric=True` to silence this warning and adopt the future behavior now.
  df_orders["order_timestamp_local"].describe()


count                               492138
unique                              485946
top              2019-06-13 11:50:00+00:00
freq                                    45
first     2019-06-01 00:00:15.369000+00:00
last      2019-12-31 23:31:59.972000+00:00
Name: order_timestamp_local, dtype: object

In [ ]:
df_orders.agg({
    "order_timestamp_local":["min","max"]
})

order_timestamp_local
min  2019-06-01T00:00:15.369Z
max  2019-12-31T23:31:59.972Z

In [ ]:
max= df_orders["order_timestamp_local"].max()
min= df_orders["order_timestamp_local"].min()

In [ ]:
max

Timestamp('2019-12-31 23:31:59.972000+0000', tz='UTC')

In [ ]:
min

Timestamp('2019-06-01 00:00:15.369000+0000', tz='UTC')

In [ ]:
df_orders[df_orders["order_timestamp_local"]==df_orders["order_timestamp_local"].min()]

order_id  order_number  \
482261  afe5e13a-d74b-4e6e-9ffa-4e7eaecf6ae4     667742584   

                  order_timestamp_local   order_shift  \
482261 2019-06-01 00:00:15.369000+00:00  weekend dawn   

          last_status_date_local  order_total  credit  paid_amount  \
482261  2019-06-01T05:06:10.898Z         25.5     4.0         21.5   

       delivery_type  scheduled  ... distance_merchant_customer  \
482261      DELIVERY      False  ...                1809.238549   

       promo_is_promotion normal_items_quantity promo_items_quantity  \
482261                0.0                   1.0                  0.0   

       order_lag_at_login order_lead_at_login          order_date_local  \
482261                1.0                 7.0  2019-06-01T00:00:00.000Z   

       valid_order                            session_id  \
482261           1  4436d781-9283-42f1-adb1-26678f274041   

                                              customer_id  
482261  a69547d1ebd59e5cd682de429b17c0ac441caf7a2cc608...  

[1 rows x 43 columns]

In [ ]:
 df_orders.query(f"order_timestamp_local == '{max}' | order_timestamp_local == '{min}'").filter(["order_id"])

order_id
176604  71b5ce4c-6ca5-4f96-9d85-ed6a2cdac4fe
482261  afe5e13a-d74b-4e6e-9ffa-4e7eaecf6ae4

In [ ]:
salvar

order_id  order_number  \
176604  71b5ce4c-6ca5-4f96-9d85-ed6a2cdac4fe    1373960993   
482261  afe5e13a-d74b-4e6e-9ffa-4e7eaecf6ae4     667742584   

                  order_timestamp_local     order_shift  \
176604 2019-12-31 23:31:59.972000+00:00  weekday dinner   
482261 2019-06-01 00:00:15.369000+00:00    weekend dawn   

          last_status_date_local  order_total  credit  paid_amount  \
176604  2020-01-01T09:08:39.856Z         80.0     4.0         76.0   
482261  2019-06-01T05:06:10.898Z         25.5     4.0         21.5   

       delivery_type  scheduled  ... distance_merchant_customer  \
176604      DELIVERY      False  ...                 879.872191   
482261      DELIVERY      False  ...                1809.238549   

       promo_is_promotion normal_items_quantity promo_items_quantity  \
176604                0.0                  12.0                  0.0   
482261                0.0                   1.0                  0.0   

       order_lag_at_login order_lead_at_login          order_date_local  \
176604                0.0                 1.0  2019-12-31T00:00:00.000Z   
482261                1.0                 7.0  2019-06-01T00:00:00.000Z   

       valid_order                            session_id  \
176604           1  b93b537a-0434-4367-9371-0ad01f930d73   
482261           1  4436d781-9283-42f1-adb1-26678f274041   

                                              customer_id  
176604  0a5aec9a37c47450108d515d1e9b46cf5bbf08cbbfeeba...  
482261  a69547d1ebd59e5cd682de429b17c0ac441caf7a2cc608...  

[2 rows x 43 columns]

In [ ]:
coluna= salvar.filter(["order_id"])

order_id
176604  71b5ce4c-6ca5-4f96-9d85-ed6a2cdac4fe
482261  afe5e13a-d74b-4e6e-9ffa-4e7eaecf6ae4

**`Q4`** Quantos usuários únicos existem na base?

In [ ]:
df_orders["customer_id"]

0         48b6d81407646ca61163f064a426ce38ac08bed2ce84f0...
1         b3f74186880f5f9a79e0865ad8beab47e22b785b8b3bc0...
2         c4f8999d4cb83cac8134b2051ffc611d5d660c049c3744...
3         ac89a3eb35853705a09d54f1e5988ad876628b986045cd...
4         fdef6bf0530b1f6507f175167aa442ade79b6e6fa085c9...
                                ...                        
492133    c9155b72411d95712a46efb2aaa99bcca59fe1d3388c89...
492134    e0c9002cf5cb0cb7d82bf3aa45693dec3a0af15f79be4e...
492135    6f3bcdd94a18419074813dd691aaeeff9458451e469fac...
492136    e3108a6e4e8e73e6856e32f45ed462720514a87378305b...
492137    5ad8475a3bd9091a740e14eac7695276b8e86cbb1c20d7...
Name: customer_id, Length: 492138, dtype: object

In [ ]:
(

)

In [ ]:
x=["helio",1111,True]

In [ ]:
lista [] tupla () set {} dicionario{key:value}

In [ ]:
dic = {"marca":"ford", "cor":"rosa", "ano":1960}

In [ ]:
dic = {"marca":["ford","byd","cao","xxx"], "cor":"rosa", "ano":1960}

In [ ]:
dic["marca"]

'ford'

In [ ]:
x=(1,2,3,45)

In [ ]:
len(set(df_orders["customer_id"]))

30079

In [ ]:
len(se(df_orders)

492138

In [ ]:
df_orders['customer_id'].nunique()

30079

## Criação da ABT

Será que variáveis como Recência, Frequência e Valor podem nos ajudar a prever se um dado cliente irá realizar um novo pedido no próximo mês?

Vamos criar variáveis que descrevem o comportamento do cliente com base em 1 mês e usar essas variáveis para prever se o cliente irá comprar no próximo mês (caso em que acontecerá um churn e talvez seja interessante oferecermos um cupom de desconto através de notificações no app).

**`Q5`** Crie um DataFrame chamado `df_features_train` que deverá conter os dados do período 2019-06-01 (inclusivo) até 2019-07-01 (exclusivo), ou seja deve conter somente os dados referente ao mês 6.

In [ ]:
df_features_train  = df_orders.query("order_timestamp_local >='2019-06-01'& order_timestamp_local < '2019-07-01'")

In [ ]:
df_features_train

order_id  order_number  \
2       e8c55557-81d4-4159-bd2b-c144e5fbabe3     694742752   
3       34a60e0d-0042-49af-9aa8-4dc526a998c2     679245732   
5       0d8b3571-be59-48c0-bb92-b7f3ce2793a0     686714753   
13      4511884b-ac72-44c2-9321-eb59e93dca45     690050151   
18      6f1d4815-206d-419c-8339-cef30c5f55d6     723621544   
...                                      ...           ...   
492095  d53be8b2-a7d9-4244-8baa-461997933335     699883360   
492098  b787d0f2-5ccf-472f-83b0-30cea4a1da75     717405884   
492099  9d1bf80d-6a37-4f8d-9a51-9b9bfdee6911     739456785   
492120  e4c20e96-28a5-47f0-948c-6c8415e63517     671873884   
492121  4ec228a6-4562-4761-acc2-262e80cb33f5     712041634   

           order_timestamp_local     order_shift    last_status_date_local  \
2       2019-06-11T20:42:32.058Z  weekday dinner  2019-06-11T22:42:54.005Z   
3       2019-06-05T19:16:42.111Z  weekday dinner  2019-06-05T21:17:13.264Z   
5       2019-06-08T14:44:02.062Z   weekend lunch  2019-06-08T15:25:13.583Z   
13      2019-06-10T00:41:14.558Z    weekend dawn  2019-06-10T06:48:56.869Z   
18      2019-06-22T16:02:59.896Z   weekend snack  2019-06-22T16:16:00.950Z   
...                          ...             ...                       ...   
492095  2019-06-13T20:28:46.634Z  weekday dinner  2019-06-13T22:29:25.090Z   
492098  2019-06-20T15:38:11.497Z   weekday snack  2019-06-20T17:38:54.073Z   
492099  2019-06-28T21:35:57.955Z  weekend dinner  2019-06-29T02:14:46.204Z   
492120  2019-06-02T13:04:07.270Z   weekend lunch  2019-06-02T15:05:31.103Z   
492121  2019-06-18T12:51:34.808Z   weekday lunch  2019-06-18T14:52:11.527Z   

        order_total  credit  paid_amount delivery_type  scheduled  ...  \
2              53.8     4.0         49.8      DELIVERY      False  ...   
3              58.9     4.0         54.9      DELIVERY      False  ...   
5             131.9     4.0        127.9      DELIVERY      False  ...   
13             46.8     4.0         42.8      DELIVERY      False  ...   
18             28.9     4.0         24.9      DELIVERY      False  ...   
...             ...     ...          ...           ...        ...  ...   
492095         81.9     4.0         77.9      DELIVERY      False  ...   
492098         29.0     4.0         25.0      DELIVERY      False  ...   
492099         92.6     4.0         88.6      DELIVERY      False  ...   
492120        191.1     4.0        187.1      DELIVERY      False  ...   
492121         33.0     4.0         29.0      DELIVERY      False  ...   

       distance_merchant_customer promo_is_promotion normal_items_quantity  \
2                     1642.258522                0.0                   2.0   
3                      441.677507                0.0                   1.0   
5                     5891.546259                0.0                  18.0   
13                    1781.144782                1.0                   1.0   
18                    2577.909442                1.0                   0.0   
...                           ...                ...                   ...   
492095                 356.568212                0.0                   1.0   
492098                 489.963218                0.0                   1.0   
492099                1769.116965                0.0                   9.0   
492120                1080.593993                0.0                   2.0   
492121                1493.210699                0.0                   1.0   

       promo_items_quantity order_lag_at_login order_lead_at_login  \
2                       0.0                5.0                 1.0   
3                       0.0                9.0                 2.0   
5                       0.0                2.0                 3.0   
13                      1.0                2.0                10.0   
18                      1.0                4.0                 1.0   
...                     ...                ...                 ...   
492095                  0.0           

**`Q6`** Crie um DataFrame chamado `df_target_train` que deverá conter os dados do período 2019-07-01 (inclusivo) até 2019-08-01 (exclusivo), ou seja deve conter somente os dados referente ao mês 7.

Depois selecione apenas a coluna `customer_id` e remova os duplicados.

**`Q7`** Faça o agrupamento no DataFrame `df_features_train` pela feature `customer_id`. Faça as agregações abaixo, renomeando o resultado para as seguintes colunas `uf`, `receita_1m`, `qtde_pedidos_1m` e `data_ultima_venda`, respectivamente.

- `customer_state_label` - `first`
- `paid_amount` - `sum`
- `order_id` - `nunique`
- `order_timestamp_local` - `max`

Lembre-se de utilizar o conceito que vimos em sala para renomear o resultado da agregação.

Também utilize o método `reset_index` para reinicar os indices.

Salve na variável `df_abt_train`.

In [ ]:
df_abt_train =(
    df_features_train
    .groupby("customer_id")
    .agg(uf = ('customer_state_label','first'),
         receita_1m = ("paid_amount","sum"),
         qtde_pedidos_1m = ("order_id","nunique"),
         data_ultima_venda =("order_timestamp_local" ,"max")
    )
    .reset_index()
)

In [ ]:
df_abt_train.head()

customer_id  uf  receita_1m  \
0  0001a8e61d8b08ad436e8e6f4adeb399b88df962c72d9d...  PR       102.0   
1  0001a9f97d01d2696cf70c7657ee2d039388d691720ff9...  SP        52.0   
2  0004720dc16aed1f98fd79f59736170e0d686199cd9ae5...  SP        59.3   
3  0006a32816a3af172048de7db87c97c4c8c7ad7e6385fa...  CE        95.2   
4  00081913eb21cd12aecc831bda704f8c6482723b55e664...  MS        27.0   

   qtde_pedidos_1m         data_ultima_venda  
0                2  2019-06-15T21:57:51.803Z  
1                1  2019-06-11T12:10:33.822Z  
2                1  2019-06-01T12:45:14.808Z  
3                2  2019-06-29T02:56:33.535Z  
4                1  2019-06-17T19:41:34.313Z

**`Q8`** Utilize o método `.assign` do DataFrame para fazer as seguintes alterações:

- Altere o tipo da coluna `data_ultima_venda` para o formato de data utilizando o comando `pd.to_datetime` com o parâmetro `utc=True`.
- Crie uma nova coluna chamada `data_ref` com o valor de `2019-07-01` e faça a converção utilizando o comando `pd.to_datetime` com o parâmetro `utc=True`
- Crie uma nova coluna chamada `recencia` e salve os dias referentes a subtração da coluna `data_ref` e `data_ultima_venda`. Lembre-se de extrair os dias (`.dt.days`)
- Por fim, selecione apenas as seguintes colunas: `['data_ref', 'customer_id', 'receita_1m', 'qtd_pedidos_1m', 'recencia']`

O resultado dessas alterações devem ser salvas na variável `df_abt_train`


**`Q9`** Faça o `merge` com o DataFrame `df_abt_train` com o `df_target_train` com os seguintes parametros:

- `how='left'`
- `on='customer_id'`
- `indicator=True`

Salve o merge na variável `df_abt_train`.


**`Q10`** Utilize o método `.assign` para criar a coluna `churn_next_month`.

Utilize a função `np.where` para fazer a criação. Compare as colunas `_merge` com `left_only`, se ambas forem iguais, o resultado deve ser 1, caso contrário deve ser 0.

* Obs: não esquecer de dropar a coluna `_merge` depois de ter criado a variável target.

In [ ]:
df_abt_train.assign(churn_next_month = lambda df: np.where(df["_merge"]=="left_only",1,0)).drop("_merge",axis=1)

**`Q11`** Verifique a distribuição da coluna `churn`.

<img align="left" width="20" src="http://dinomagri.com/imgs/checkpoint.png" /> &nbsp; **CHECKPOINT** - O DataFrame `df_abt_train` deverá conter exatamente `(30079, 7)`. E a distribuição da variável target deve ser `0.613086` para não-churn (0) e `0.386914` para churn (1).

**`Q12`** Crie uma ABT Out of Time avançando 1 mês no código anterior. Porque devemos criar uma ABT Out of Time?

Utilize o mês 08 para criar essa ABT.

<img align="left" width="20" src="http://dinomagri.com/imgs/checkpoint.png" /> &nbsp; **CHECKPOINT** - O DataFrame `df_abt_oot` deverá conter exatamente `(18441, 7)`. E a distribuição da variável target deve ser `0.759558` para não-churn (0) e `0.240442` para churn (1).

**`Q13`** Faça um estudo de valores faltantes tanto na `df_abt_train` e `df_abt_oot`.

**`Q14`** Salve os DataFrames ( `df_abt_train` e `df_abt_oot`) em arquivos CSVs para serem utiliza-los posteriormente.

O nome do arquivo deve ser:

*   `/content/drive/MyDrive/datasets/ifood/abt_classificacao_ifood_train.csv`
*   `/content/drive/MyDrive/datasets/ifood/abt_classificacao_ifood_oot.csv`
